events/diskpersist.go

```diff
@@ -1,19 +1,710 @@
+	if len(p.evtbuf) == 0 {
+		return nil
+	}
+
+	_, err := io.Copy(p.logfi, p.outbuf)
+	if err != nil {
+		return err
+	}
+
+	p.outbuf.Truncate(0)
+
+	for _, ej := range p.evtbuf {
+		p.broadcast(ej.Evt)
+		ej.Buffer.Truncate(0)
+		p.buffers.Put(ej.Buffer)
+	}
+
+	p.evtbuf = p.evtbuf[:0]
+
+	return nil
+}
+
+func (p *DiskPersistence) doPersist(j persistJob) error {
+	b := j.Buf
+	e := j.Evt
+	seq := p.curSeq
+	p.curSeq++
+
+	binary.LittleEndian.PutUint64(b[20:], uint64(seq))
+
+	switch {
+	case e.RepoCommit != nil:
+		e.RepoCommit.Seq = seq
+	case e.RepoHandle != nil:
+		e.RepoHandle.Seq = seq
+	default:
+		// only those two get peristed right now
+		// we shouldnt actually ever get here...
+		return nil
+	}
+
+	// TODO: does this guarantee a full write?
+	_, err := p.outbuf.Write(b)
+	if err != nil {
+		return err
+	}
+
+	p.evtbuf = append(p.evtbuf, j)
+
+	if seq%p.eventsPerFile == 0 {
+		if err := p.flushLog(context.TODO()); err != nil {
+			return err
+		}
+
+		// time to roll the log file
+		if err := p.swapLog(context.TODO()); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *DiskPersistence) Persist(ctx context.Context, e *XRPCStreamEvent) error {
+	buffer := p.buffers.Get().(*bytes.Buffer)
+
+	buffer.Truncate(0)
+
+	buffer.Write(emptyHeader)
+
+	var did string
+	var evtKind uint32
+	switch {
+	case e.RepoCommit != nil:
+		evtKind = evtKindCommit
+		did = e.RepoCommit.Repo
+		if err := e.RepoCommit.MarshalCBOR(buffer); err != nil {
+			return fmt.Errorf("failed to marshal: %w", err)
+		}
+	case e.RepoHandle != nil:
+		evtKind = evtKindHandle
+		did = e.RepoHandle.Did
+		if err := e.RepoHandle.MarshalCBOR(buffer); err != nil {
+			return fmt.Errorf("failed to marshal: %w", err)
+		}
+	default:
+		return nil
+		// only those two get peristed right now
+	}
+
+	usr, err := p.uidForDid(ctx, did)
+	if err != nil {
+		return err
+	}
+
+	b := buffer.Bytes()
+
+	binary.LittleEndian.PutUint32(b, 0)
+	binary.LittleEndian.PutUint32(b[4:], evtKind)
+	binary.LittleEndian.PutUint32(b[8:], uint32(len(b)-headerSize))
+	binary.LittleEndian.PutUint64(b[12:], uint64(usr))
+
+	return p.addJobToQueue(persistJob{
+		Buf:    b,
+		Evt:    e,
+		Buffer: buffer,
+	})
+}
+
+type evtHeader struct {
+	Flags uint32
+	Kind  uint32
+	Seq   int64
+	Usr   util.Uid
+	Len   int64
 }
 
-func NewDiskPersistence(dir string) (*DiskPersistence, error) {
-	return &DiskPersistence{
-		dir: dir,
+const headerSize = 4 + 4 + 4 + 8 + 8
+
+func readHeader(r io.Reader, scratch []byte) (*evtHeader, error) {
+	if len(scratch) < headerSize {
+		return nil, fmt.Errorf("must pass scratch buffer of at least %d bytes", headerSize)
+	}
+
+	scratch = scratch[:headerSize]
+	_, err := io.ReadFull(r, scratch)
+	if err != nil {
+		return nil, fmt.Errorf("reading header: %w", err)
+	}
+
+	flags := binary.LittleEndian.Uint32(scratch[:4])
+	kind := binary.LittleEndian.Uint32(scratch[4:8])
+	l := binary.LittleEndian.Uint32(scratch[8:12])
+	usr := binary.LittleEndian.Uint64(scratch[12:20])
```

Comment:
Same as above comment, this is fine if we set `usr` to be a fixed-width `uint64` (though in the header above it seems like you want it to be a `uint32`?), also above you have `l` here as a `uint32` but you've declared it in the `evtHeader` struct as an `int64`. Which is it? I think these short variable names don't help when trying to do byte manipulation fwiw.